syntax = "proto3";

package mc.comm;

option go_package = "mc/comm";

message Scene {
    repeated int32 nodes = 1;
}

message Node {
    int32 mesh = 1;
}

message Attributes {
    int32 POSITION = 1;
    int32 NORMAL = 2;
    int32 TEXCOORD_0 = 3;
}

message Primitive {
    int32 indices = 1;
    Attributes attributes = 2;
}

message Mesh {
    repeated Primitive primitives= 1;
}

message Texture {
    int32 sampler = 1;
    int32 source = 2;
}

message Image {
    string uri = 1;
}

message Sampler {
    int32 magFilter = 1;
    int32 minFilter = 2;
    int32 wrapS = 3;
    int32 wrapT = 4;
}

message Buffer {
    string uri = 1;
    int32 byteLength = 2;
}

message BufferView {
    int32 buffer = 1;
    int32 byteOffset = 2;
    int32 byteLength = 3;
    int32 byteStride = 4;
    int32 target = 5;
}

message Accessor {
    int32 bufferView = 1;
    int32 byteOffset = 2;
    int32 componentType = 3;
    int32 count = 4;
    string type = 5;
    repeated int32 max = 6;
    repeated int32 min = 7;
}

message Asset {
    string version = 1;
}

message GLTF {
    int32 scene = 1;
    repeated Scene scenes = 2;
    repeated Node nodes = 3;
    repeated Mesh meshes = 4;
    repeated Texture textures = 5;
    repeated Image images = 6;
    repeated Sampler samplers = 7;
    repeated Buffer buffers = 8;
    repeated BufferView bufferViews = 9;
    repeated Accessor accessors = 10;
    Asset asset = 11;
}
